File: NumberOfWaysDest.cpp
 class Solution {
public:
    
     
    int countPaths(int n, vector<vector<int>>& roads) {

          long long MOD = 1e9 + 7;
          vector<long long>distance(n, 1e15), ways(n, 0);

          distance[0] = 0;

          vector<pair<int,long long>>adj[n];

          for(auto it: roads){
              adj[it[0]].push_back({it[1], it[2]}); 
              adj[it[1]].push_back({it[0],it[2]});
          }

          priority_queue<pair<long long,int>, vector<pair<long long,int>>, greater<pair<long long,int>>>pq;

          pq.push({0,0});

          ways[0]=1;

          while(!pq.empty()){
                 
                  pair<long long,int>p = pq.top();

                  pq.pop();

                  int node = p.second;
                  long long  dist = p.first;

                  if(distance[node] < dist)continue;

                  for(auto it: adj[node]){
                        
                        int to = it.first;
                        int wt = it.second;
                        if(distance[to] > dist + wt){
                              distance[to] = dist + wt;
                              ways[to] = ways[node];
                              pq.push({distance[to],to});
                        }
                        else if(distance[to] == dist + wt){
                               ways[to] = (ways[to] + ways[node])%(MOD);
                        }
                  }
          }

          return ways[n-1]%MOD;

    }
};

File: NumberOfWaysDest.py
import heapq

class Solution:
    def countPaths(self, n, roads):
        MOD = 10**9 + 7
        distance = [float('inf')] * n
        ways = [0] * n

        distance[0] = 0
        adj = [[] for _ in range(n)]

        for it in roads:
            adj[it[0]].append((it[1], it[2]))
            adj[it[1]].append((it[0], it[2]))

        pq = [(0, 0)]
        ways[0] = 1

        while pq:
            dist, node = heapq.heappop(pq)

            if distance[node] < dist:
                continue

            for to, wt in adj[node]:
                if distance[to] > dist + wt:
                    distance[to] = dist + wt
                    ways[to] = ways[node]
                    heapq.heappush(pq, (distance[to], to))
                elif distance[to] == dist + wt:
                    ways[to] = (ways[to] + ways[node]) % MOD

        return ways[n-1] % MOD
